spec/abci++/abci++_methods_002_draft.md

```diff
@@ -305,106 +329,140 @@ title: Methods
       In particular, `ResponsePrepareProposal.tx_records` will be deemed invalid if
         * There is a duplicate transaction in the list.
-        * A new or modified transaction is marked as `UNMODIFIED` or `REMOVED`.
-        * An unmodified transaction is marked as `ADDED`.
+        * A new transaction is marked as `UNMODIFIED` or `REMOVED`.
+        * An existing transaction is marked as `ADDED`.
         * A transaction is marked as `UNKNOWN`.
-    * If Tendermint fails to validate the `ResponsePrepareProposal`, Tendermint will assume the application is faulty and crash.
+    * If Tendermint fails to validate the `ResponsePrepareProposal`, Tendermint will assume the
+      Application is faulty and crash.
     * The implementation of `PrepareProposal` can be non-deterministic.
 
-#### When does Tendermint call it?
+#### When does Tendermint call `PrepareProposal`?
 
 When a validator _p_ enters Tendermint consensus round _r_, height _h_, in which _p_ is the proposer,
 and _p_'s _validValue_ is `nil`:
 
 1. _p_'s Tendermint collects outstanding transactions from the mempool
-    * The transactions will be collected in order of priority
-    * Let $C$ the list of currently collected transactions
-    * The collection stops when any of the following conditions are met
-        * the mempool is empty
-        * the total size of transactions $\in C$ is greater than or equal to `consensusParams.block.max_bytes`
-        * the sum of `GasWanted` field of transactions $\in C$ is greater than or equal to
-          `consensusParams.block.max_gas`
+    * the transactions will be collected in order of priority
     * _p_'s Tendermint creates a block header.
-2. _p_'s Tendermint calls `RequestPrepareProposal` with the newly generated block.
-   The call is synchronous: Tendermint's execution will block until the Application returns from the call.
-3. The Application checks the block (hashes, transactions, commit info, misbehavior). Besides,
-    * in same-block execution mode, the Application can (and should) provide `ResponsePrepareProposal.app_hash`,
-      `ResponsePrepareProposal.validator_updates`, or
+2. _p_'s Tendermint calls `RequestPrepareProposal` with the newly generated block, the local
+   commit of the previous height (with vote extensions), and any outstanding evidence of
+   misbahavior. The call is synchronous: Tendermint's execution will block until the Application
+   returns from the call.
+3. The Application uses the information received (transactions, commit info, misbehavior, time) to
+    (potentially) modify the proposal.
+    * in same-block execution mode, the Application fully executes the block and provides values
+      for `ResponsePrepareProposal.app_hash`, `ResponsePrepareProposal.tx_results`,
+      `ResponsePrepareProposal.validator_updates`, and
+      `ResponsePrepareProposal.consensus_param_updates`.
+    * in next-block execution mode, _p_'s Tendermint will ignore the values for
```

Comment:
same-block next block still referenced here?

Reply:
Yes. Next PR is removing it, but only from v0.36.x